webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala

```diff
@@ -541,41 +566,45 @@ object ConstructResponseUtilV2 {
-          * @return the same resource, with any nested resources attached to it.
-          */
+         * Given a resource IRI, finds any link values in the resource, and recursively embeds the target resource in each link value.
+         *
+         * @param resourceIri      the IRI of the resource to start with.
+         * @param alreadyTraversed a set (initially empty) of the IRIs of resources that this function has already
+         *                         traversed, to prevent an infinite loop if a cycle is encountered.
+         * @return the same resource, with any nested resources attached to it.
+         */
         def nestResources(resourceIri: IRI, alreadyTraversed: Set[IRI] = Set.empty[IRI]): ResourceWithValueRdfData = {
-            val resource = flatResourcesWithValues(resourceIri)
+            val resource = visibleResources(resourceIri)
 
             val transformedValuePropertyAssertions: RdfPropertyValues = resource.valuePropertyAssertions.map {
                 case (propIri, values) =>
-                    val transformedValues = values.map {
-                        value =>
+                    val transformedValues: Seq[ValueRdfData] = values.foldLeft(Vector.empty[ValueRdfData]) {
+                        case (acc: Vector[ValueRdfData], value: ValueRdfData) =>
                             if (value.valueObjectClass.toString == OntologyConstants.KnoraBase.LinkValue) {
                                 val dependentResourceIri: IRI = value.requireIriObject(OntologyConstants.Rdf.Object.toSmartIri)
 
                                 if (alreadyTraversed(dependentResourceIri)) {
-                                    value
+                                    acc :+ value
                                 } else {
-                                    // If we don't have the dependent resource, that means that the user doesn't have
-                                    // permission to see it, or it's been marked as deleted. Just return the link
-                                    // value without a nested resource.
-                                    if (flatResourcesWithValues.contains(dependentResourceIri)) {
+                                    // Do we have the dependent resource?
+                                    if (dependentResourceIrisVisible.contains(dependentResourceIri)) {
+                                        // Yes. Nest it in the link value.
                                         val dependentResource: ResourceWithValueRdfData = nestResources(dependentResourceIri, alreadyTraversed + resourceIri)
 
-                                        value.copy(
+                                        acc :+ value.copy(
                                             nestedResource = Some(dependentResource)
                                         )
+                                    } else if (dependentResourceIrisNotVisible.contains(dependentResourceIri)) {
```

Comment:
> now if the dependant resource is forbidden (i.e. user does not have permission to see), then it is skipped. If the dependant resource is deleted, the link value is however returned. Have I understood it right?

Yes. Do you think this makes sense? Or should we do the same thing in both cases?

> If so, when a user requests a resource, the link value property itself will not be visible to the user either if the user does not have permission to see the dependent resource, right?

Right, it should not be visible. I will check and see if that's what really happens.

Reply:
It turns out that the link property was not filtered out if we filtered out all its values. I've fixed this and added a test in 35fafe4. Good catch, thanks!

Reply:
We have to return a `LinkValue` that points to a deleted resource, so the user can change the link to point to a different resource. In the case of a `LinkValue` that points to a resource that the user doesn't have permission to see, I filtered it out because @loicjaouen [said he preferred it that way](https://discuss.dasch.swiss/t/how-to-deal-with-forbidden-resources-and-values-consistently/153/2?u=benjamingeer). I think it makes sense because there isn't anything that the user could do with that link.

Reply:
> Yes. Do you think this makes sense? Or should we do the same thing in both cases?
yes, I believe it makes sense as it is now. In this way, the difference between a deleted dependant resource and a forbidden one is clear and the user will not be confused by seeing a link value but not its dependant resource.

Reply:
OK great. In b990663 I've added a test that gets a link to a deleted resource, to check that the link value is returned correctly.

Reply:
Looks good to me!